### PR TITLE
fuse: implement the FUSE_UNLINK command

### PR DESCRIPTION
**Pull request by boyuan-he**

### CODE REVIEW

File: pkg/abi/linux/fuse.go [hunk 1]
```diff
@@ -729,3 +729,25 @@ func (r *FUSEDirent) UnmarshalBytes(src []byte) {
...
+//
+// Dynamically-sized objects cannot be marshalled.
+type FUSEUnlinkIn struct {
+	marshal.StubMarshallable
+
+	// Name of the node to unlink.
+	Name string
+}
+
+// MarshalBytes serializes r.name to the dst buffer.
```
Comment (jinmouil) [inline pkg/abi/linux/fuse.go]: `, assuming len(buf) == len(r.name) + 1 && buf[len(r.name)] == '\0'`

File: pkg/sentry/fsimpl/fuse/fusefs.go [hunk 2]
```diff
@@ -428,6 +428,29 @@ func (i *inode) NewSymlink(ctx context.Context, name, target string) (*vfs.Dentr
...
+	in := linux.FUSEUnlinkIn{Name: name}
+	req, err := i.fs.conn.NewRequest(auth.CredentialsFromContext(ctx), uint32(kernelTask.ThreadID()), i.NodeID, linux.FUSE_UNLINK, &in)
+	if err != nil {
+		return err
+	}
+	res, err := i.fs.conn.Call(kernelTask, req)
+	if err != nil {
+		return err
+	}
+	// only error would be returned, so no need to unmarshal res.
```
Comment (jinmouil) [inline pkg/sentry/fsimpl/fuse/fusefs.go]: "Only return error. Discard res."

File: pkg/sentry/fsimpl/kernfs/filesystem.go [hunk 3]
```diff
@@ -768,6 +768,10 @@ func (fs *Filesystem) SymlinkAt(ctx context.Context, rp *vfs.ResolvingPath, targ
 func (fs *Filesystem) UnlinkAt(ctx context.Context, rp *vfs.ResolvingPath) error {
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
+
+	// Store the name before walkExistingLocked as rp will be advanced past the
+	// name in the following call.
+	name := rp.Component()
```
Comment (jinmouil) [inline pkg/sentry/fsimpl/kernfs/filesystem.go:774]: I think I saw this in a previous change as well. Is it correct for all fs using kernfs?

File: pkg/sentry/fsimpl/kernfs/kernfs.go [hunk 4]
```diff
@@ -267,6 +268,36 @@ func (d *Dentry) InsertChildLocked(name string, child *Dentry) {
 	d.children[name] = child
 }
 
+// RemoveChild removes child from the vfs dentry cache. This does not update the
+// directory inode or modify the inode to be unlinked. so calling this on  its own
```
Comment (jinmouil) [inline pkg/sentry/fsimpl/kernfs/kernfs.go]: `So ... on its own`

File: pkg/sentry/fsimpl/kernfs/kernfs.go [hunk 5]
```diff
@@ -267,6 +268,36 @@ func (d *Dentry) InsertChildLocked(name string, child *Dentry) {
...
+// isn't sufficient to remove a child from a directory.
+//
+// Precondition: d must represent a directory inode.
+func (d *Dentry) RemoveChild(name string, child *vfs.Dentry) error {
+	d.dirMu.Lock()
+	defer d.dirMu.Unlock()
+	return d.RemoveChildLocked(name, child)
+}
+
+// RemoveChildLocked is equivalent to RemoveChild, with additional
```
Comment (jinmouil) [inline pkg/sentry/fsimpl/kernfs/kernfs.go:282]: They look definitely not "equivalent" to me

File: test/fuse/linux/unlink_test.cc [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  SetServerResponse(FUSE_UNLINK, iov_out);
+
+  ASSERT_THAT(unlink(test_file_path.c_str()), SyscallSucceeds());
+  struct fuse_in_header in_header;
+  std::vector<char> unlinked_file(test_file_.length() + 1);
+  auto iov_in = FuseGenerateIovecs(in_header, unlinked_file);
+  GetServerActualRequest(iov_in);
+  EXPECT_EQ(in_header.len, sizeof(in_header) + test_file_.length()  + 1);
+  EXPECT_EQ(in_header.opcode, FUSE_UNLINK);
+  EXPECT_EQ(0, memcmp(unlinked_file.data(), test_file_.c_str(),
```
Comment (jinmouil) [inline test/fuse/linux/unlink_test.cc]: Why not use `EXPECT_EQ(0, unlinked_file, std::vector<char>(test_file_))`

File: test/fuse/linux/unlink_test.cc [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  
+  struct fuse_out_header out_header = {
+      .len = sizeof(struct fuse_out_header),
+      .error = -ENOENT,
+  };
+  auto iov_out = FuseGenerateIovecs(out_header);
+  SetServerResponse(FUSE_UNLINK, iov_out);
+
+  ASSERT_THAT(unlink(test_file_path.c_str()), SyscallFailsWithErrno(ENOENT));
+  SkipServerActualRequest();
```
Comment (jinmouil) [inline test/fuse/linux/unlink_test.cc:77]: Why not check the request as well?

File: test/fuse/linux/unlink_test.cc [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+#include <string>
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "test/util/fuse_util.h"
+#include "test/util/test_util.h"
+
+#include "fuse_base.h"
```
Comment (craig08) [inline test/fuse/linux/unlink_test.cc]: Google's C++ style.

File: pkg/sentry/fsimpl/kernfs/filesystem.go (hunk 3 above)
Comment (boyuan-he) [reply pkg/sentry/fsimpl/kernfs/filesystem.go:774]: Yes

File: test/fuse/linux/unlink_test.cc (hunk 7 above)
Comment (boyuan-he) [reply test/fuse/linux/unlink_test.cc:77]: The request would be exactly the same as previous test.

File: pkg/sentry/fsimpl/kernfs/kernfs.go (hunk 5 above)
Comment (boyuan-he) [reply pkg/sentry/fsimpl/kernfs/kernfs.go:282]: `with additional precondition`, also this comment is slightly modified from `InsertChildLocked`

**Pull request merged**